network/tcp.go

```diff
@@ -0,0 +1,411 @@
+	var conn net.Conn
+	netAddr := addr.NetworkAddress()
+	for i := 0; i < MaxRetryConnect; i++ {
+		conn, err = net.Dial("tcp", netAddr)
+		if err != nil {
+			time.Sleep(WaitRetry)
+		} else {
+			break
+		}
+		time.Sleep(WaitRetry)
+	}
+	if conn == nil {
+		return nil, fmt.Errorf("Could not connect to %s: %s", addr, err)
+	}
+	c := TCPConn{
+		endpoint: addr,
+		conn:     conn,
+	}
+	return &c, err
+}
+
+// Receive calls the receive routine to get the bytes from the connection then
+// it decodes the buffer.
+// It returns the Packet with the Msg field decoded
+// or EmptyApplicationPacket and an error if something wrong happened.
+func (c *TCPConn) Receive() (nm Packet, e error) {
+	defer func() {
+		if err := recover(); err != nil {
+			e = fmt.Errorf("Error Received message: %v", err)
+			nm = EmptyApplicationPacket
+		}
+	}()
+
+	var am Packet
+	buff, err := c.receiveRaw()
+	if err != nil {
+		return EmptyApplicationPacket, err
+	}
+
+	err = am.UnmarshalBinary(buff)
+	if err != nil {
+		return EmptyApplicationPacket, fmt.Errorf("Error unmarshaling message type %s: %s", am.MsgType.String(), err.Error())
+	}
+	am.From = c.Remote()
+	return am, nil
+}
+
+// receiveRaw reads the size of the message, then the
+// whole message. It returns the raw message as slice of bytes.
+// If there is no message available, it blocks until one becomes
+// available.
+// In case of an error it returns a nil slice and the error.
+func (c *TCPConn) receiveRaw() ([]byte, error) {
+	c.receiveMutex.Lock()
+	defer c.receiveMutex.Unlock()
+	// First read the size
+	var total Size
+	if err := binary.Read(c.conn, globalOrder, &total); err != nil {
+		return nil, handleError(err)
+	}
+	b := make([]byte, total)
+	var read Size
+	var buffer bytes.Buffer
+	for read < total {
+		// Read the size of the next packet.
+		n, err := c.conn.Read(b)
+		// Quit if there is an error.
+		if err != nil {
+			return nil, handleError(err)
+		}
+		// Append the read bytes into the buffer.
+		if _, err := buffer.Write(b[:n]); err != nil {
+			log.Error("Couldn't write to buffer:", err)
+		}
+		read += Size(n)
+		b = b[n:]
+	}
+
+	// register how many bytes we read.
+	c.updateRx(uint64(read))
+	return buffer.Bytes(), nil
+}
+
+// Send converts the NetworkMessage into an ApplicationMessage
+// and sends it using send().
+// It returns an error if anything was wrong.
+func (c *TCPConn) Send(obj Body) error {
+	c.sendMutex.Lock()
+	defer c.sendMutex.Unlock()
+	am, err := NewNetworkPacket(obj)
+	if err != nil {
+		return fmt.Errorf("Error converting packet: %v\n", err)
+	}
+	log.Lvlf5("Message SEND => %+v", am)
+	var b []byte
+	b, err = am.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("Error marshaling  message: %s", err.Error())
+	}
+	return c.sendRaw(b)
+}
+
+// sendRaw writes the number of bytes of the message to the network then the
+// whole message b in slices of size maxChunkSize.
+// In case of an error it aborts and returns error.
+func (c *TCPConn) sendRaw(b []byte) error {
```

Comment:
Same as above: Should be fully tested (also everything that could go wrong). There should be a `TestsendRaw` even if it is an unexported method. It's too important...